pkg/reconciler/pipelinerun/resources/apply_test.go

```diff
@@ -2034,6 +2034,235 @@ func TestApplyParameters_ArrayIndexing(t *testing.T) {
 	}
 }
 
+func TestApplyReplacementsMatrix(t *testing.T) {
+	ctx := context.Background()
+	cfg := config.FromContextOrDefaults(ctx)
+	cfg.FeatureFlags.EnableAPIFields = config.BetaAPIFields
+	ctx = config.ToContext(ctx, cfg)
+	for _, tt := range []struct {
+		name     string
+		original v1beta1.PipelineSpec
+		params   []v1beta1.Param
+		expected v1beta1.PipelineSpec
+	}{{
+		name: "matrix params replacement",
+		original: v1beta1.PipelineSpec{
+			Params: []v1beta1.ParamSpec{{
+				Name: "first-param", Type: v1beta1.ParamTypeArray, Default: v1beta1.NewStructuredValues("default-value", "default-value-again"),
+			}, {
+				Name: "second-param", Type: v1beta1.ParamTypeString,
+			}},
+			Tasks: []v1beta1.PipelineTask{{
+				Matrix: &v1beta1.Matrix{
+					Params: v1beta1.Params{{
+						Name: "first-param", Value: *v1beta1.NewStructuredValues("$(params.first-param[1])"),
+					}, {
+						Name: "second-param", Value: *v1beta1.NewStructuredValues("$(params.second-param[0])"),
+					}}},
+			}},
+		},
+		params: v1beta1.Params{{Name: "second-param", Value: *v1beta1.NewStructuredValues("second-value", "second-value-again")}},
+		expected: v1beta1.PipelineSpec{
+			Params: []v1beta1.ParamSpec{{
+				Name: "first-param", Type: v1beta1.ParamTypeArray, Default: v1beta1.NewStructuredValues("default-value", "default-value-again"),
+			}, {
+				Name: "second-param", Type: v1beta1.ParamTypeString,
+			}},
+			Tasks: []v1beta1.PipelineTask{{
+				Matrix: &v1beta1.Matrix{
+					Params: v1beta1.Params{{
+						Name: "first-param", Value: *v1beta1.NewStructuredValues("default-value-again"),
+					}, {
+						Name: "second-param", Value: *v1beta1.NewStructuredValues("second-value"),
+					}}},
+			}},
+		},
+	}, {
+		name: "matrix include replacement",
+		original: v1beta1.PipelineSpec{
+			Params: []v1beta1.ParamSpec{{
+				Name: "first-param", Type: v1beta1.ParamTypeArray, Default: v1beta1.NewStructuredValues("default-value", "default-value-again"),
+			}, {
+				Name: "second-param", Type: v1beta1.ParamTypeString,
+			},
+			},
+			Tasks: []v1beta1.PipelineTask{{
+				Matrix: &v1beta1.Matrix{
+					Include: []v1beta1.MatrixInclude{{
+						Name: "build-1",
+						Params: v1beta1.Params{{
+							Name: "first-param", Value: *v1beta1.NewStructuredValues("$(params.first-param[1])"),
+						}, {
+							Name: "second-param", Value: *v1beta1.NewStructuredValues("$(params.second-param[0])"),
+						}}},
+					}},
+			}},
+		},
+		params: v1beta1.Params{{Name: "second-param", Value: *v1beta1.NewStructuredValues("second-value", "second-value-again")}},
+		expected: v1beta1.PipelineSpec{
+			Params: []v1beta1.ParamSpec{{
+				Name: "first-param", Type: v1beta1.ParamTypeArray, Default: v1beta1.NewStructuredValues("default-value", "default-value-again"),
```

Comment:
it might also be useful to test a case where defaults are overridden

Reply:
Added a test case for overriding defaults